File: parser.py
"""Parse Dirty South Trivia .docx files into structured JSON."""

import json
import re
from dataclasses import dataclass, field, asdict
from pathlib import Path

from docx import Document


@dataclass
class Question:
    number: int
    text: str
    answer: str
    choices: list[str] = field(default_factory=list)  # for multiple choice
    is_internet_only: bool = False


@dataclass
class Round:
    number: int
    title: str  # e.g., "Random", "Zero Proof"
    round_type: str  # "standard", "video", "progressive", "list"
    points_per_question: int = 1
    joker_eligible: bool = True
    theme_description: str = ""
    questions: list[Question] = field(default_factory=list)
    # For progressive reveal rounds (Guess Who/Where)
    clues: list[dict] = field(default_factory=list)  # [{points: 10, text: "..."}, ...]
    progressive_answer: str = ""


@dataclass
class MiniGame:
    number: int
    title: str
    description: str


@dataclass
class Quiz:
    quiz_number: int
    date: str
    rounds: list[Round] = field(default_factory=list)
    mini_games: list[MiniGame] = field(default_factory=list)
    tie_breaker_question: str = ""
    tie_breaker_answer: str = ""

    def to_dict(self):
        return asdict(self)


def parse_docx(path: str | Path) -> Quiz:
    """Parse a Dirty South Trivia .docx file into a Quiz object."""
    doc = Document(str(path))
    paragraphs = [p.text.strip() for p in doc.paragraphs]

    # Extract quiz number and date from first non-empty line
    quiz_number = 0
    date = ""
    for p in paragraphs:
        if m := re.match(r"Pub Quiz (\d+)\s*[–-]\s*(.+)", p):
            quiz_number = int(m.group(1))
            date = m.group(2).strip()
            break

    quiz = Quiz(quiz_number=quiz_number, date=date)

    # Find all round starts, mini games, and tie breaker
    sections = []
    for i, p in enumerate(paragraphs):
        # Round header (the detailed one, not the overview list)
        # Match "Round N - Title" that appears AFTER the overview section
        if rm := re.match(r"Round\s+(\d+)\s*[–-]\s*(.+)", p):
            sections.append(("round", i, int(rm.group(1)), rm.group(2).strip()))
        elif re.match(r"Mini Game \d+", p, re.IGNORECASE):
            sections.append(("minigame", i, 0, p))
        elif re.match(r"Tie\s*Breaker", p, re.IGNORECASE):
            sections.append(("tiebreaker", i, 0, ""))

    # Deduplicate rounds — the overview lists rounds briefly, then they appear in detail
    # Keep only the LAST occurrence of each round number (the detailed one)
    seen_rounds = {}
    for s in sections:
        if s[0] == "round":
            seen_rounds[s[2]] = s
    deduped = []
    used_round_nums = set()
    for s in sections:
        if s[0] == "round":
            if s[2] not in used_round_nums:
                # Skip overview entries — use the last one
                used_round_nums.add(s[2])
                continue
            deduped.append(seen_rounds[s[2]])
        else:
            deduped.append(s)
    # Actually simpler: just take sections where the round has content after it
    # Deduplicate rounds — keep only the LAST occurrence of each round number
    # (the overview lists them first briefly, then the detailed version appears later)
    last_round_idx = {}  # round_num -> index in sections
    for i, s in enumerate(sections):
        if s[0] == "round":
            last_round_idx[s[2]] = i

    sections = [
        s for i, s in enumerate(sections)
        if s[0] != "round" or last_round_idx.get(s[2]) == i
    ]

    # Process each section
    for idx, section in enumerate(sections):
        # Determine section boundaries
        start_line = section[1]
        if idx + 1 < len(sections):
            end_line = sections[idx + 1][1]
        else:
            end_line = len(paragraphs)

        section_text = paragraphs[start_line:end_line]

        if section[0] == "round":
            round_obj = _parse_round(section[2], section[3], section_text)
            quiz.rounds.append(round_obj)
        elif section[0] == "minigame":
            mg = _parse_mini_game(section_text)
            quiz.mini_games.append(mg)
        elif section[0] == "tiebreaker":
            q, a = _parse_tie_breaker(section_text)
            quiz.tie_breaker_question = q
            quiz.tie_breaker_answer = a

    return quiz


def _parse_round(round_num: int, title: str, lines: list[str]) -> Round:
    """Parse a round section into a Round object."""
    # Detect round type
    is_video = "video" in title.lower() or any("video" in l.lower() and "montage" in l.lower() for l in lines[:5])
    is_progressive = any(re.match(r"\d+\s+POINTS?:", l) for l in lines)

    # Scoring
    points = 1
    joker = True
    description = ""

    for l in lines[:10]:
        if m := re.match(r"(\d+)\s+Points?\s+Per\s+Correct\s+Answer", l, re.IGNORECASE):
            points = int(m.group(1))
        if "joker is not in play" in l.lower() or "no joker" in l.lower():
            joker = False
        # Theme description — lines that aren't round headers, scoring, or questions
        if (l and not re.match(r"Round\s+\d+", l) and
            "point" not in l.lower() and "joker" not in l.lower() and
            not re.match(r"\d+\.", l) and
            "video montage" not in l.lower() and
            "youtube" not in l.lower()):
            if not description and len(l) > 20:
                description = l

    if is_progressive:
        return _parse_progressive_round(round_num, title, lines, points, joker, description)

    round_type = "video" if is_video else "standard"
    rnd = Round(
        number=round_num, title=title, round_type=round_type,
        points_per_question=points, joker_eligible=joker,
        theme_description=description,
    )

    # Extract questions and answers
    i = 0
    while i < len(lines):
        line = lines[i]
        # Match question lines: "N. question text" or "N.  question text"
        qm = re.match(r"(\d+)\.\s+(.+)", line)
        if qm:
            q_num = int(qm.group(1))
            q_text = qm.group(2).strip()

            is_internet = "internet" in q_text.lower() and "question" in q_text.lower()

            # Look ahead for choices (A. B. C. D.) or answer
            choices = []
            answer = ""
            j = i + 1
            while j < len(lines):
                next_line = lines[j]
                if not next_line:
                    j += 1
                    continue
                # Check if it's a choice
                if re.match(r"[A-D]\.\s+", next_line):
                    choices.append(next_line)
                    if "correct" in next_line.lower() or "✅" in next_line:
                        # Extract answer from correct choice
                        answer = re.sub(r"[A-D]\.\s+", "", next_line)
                        answer = re.sub(r"\s*\(?\s*✅?\s*Correct\s*\)?\s*", "", answer, flags=re.IGNORECASE).strip()
                    j += 1
                elif re.match(r"\d+\.\s+", next_line):
                    # Next question — stop
                    break
                elif next_line.startswith("Please turn") or next_line.startswith("Round "):
                    break
                else:
                    if not answer and not re.match(r"[A-D]\.\s+", next_line):
                        answer = next_line
                    j += 1
                    break

            # For multiple choice without explicit correct marker, answer is after choices
            if choices and not answer:
                # Look for answer after choices
                while j < len(lines) and not lines[j]:
                    j += 1
                if j < len(lines) and not re.match(r"\d+\.\s+", lines[j]):
                    answer = lines[j]

            rnd.questions.append(Question(
                number=q_num, text=q_text, answer=answer,
                choices=choices, is_internet_only=is_internet,
            ))
            i = j
        else:
            i += 1

    return rnd


def _parse_progressive_round(round_num, title, lines, points, joker, description) -> Round:
    """Parse a Guess Who/Where progressive reveal round."""
    rnd = Round(
        number=round_num, title=title, round_type="progressive",
        points_per_question=0, joker_eligible=joker,
        theme_description=description,
    )

    for i, line in enumerate(lines):
        if m := re.match(r"(\d+)\s+POINTS?:\s*(.+)", line):
            rnd.clues.append({"points": int(m.group(1)), "text": m.group(2).strip()})
        elif line.startswith("ANSWER:"):
            rnd.progressive_answer = line.replace("ANSWER:", "").strip()

    # If no explicit ANSWER line, look for standalone answer after clues
    if not rnd.progressive_answer and rnd.clues:
        # Find the line after the last clue
        last_clue_idx = 0
        for i, line in enumerate(lines):
            if re.match(r"\d+\s+POINTS?:", line):
                last_clue_idx = i
        for j in range(last_clue_idx + 1, len(lines)):
            if lines[j] and not lines[j].startswith("Please") and "pens" not in lines[j].lower():
                rnd.progressive_answer = lines[j]
                break

    return rnd


def _parse_mini_game(lines: list[str]) -> MiniGame:
    """Parse a mini game section."""
    title = lines[0] if lines else ""
    # Extract the mini game number
    num_match = re.match(r"Mini Game (\d+)", title, re.IGNORECASE)
    num = int(num_match.group(1)) if num_match else 0
    # Clean title
    title = re.sub(r"Mini Game \d+\s*[–-]\s*", "", title).strip()
    desc = "\n".join(l for l in lines[1:] if l)
    return MiniGame(number=num, title=title, description=desc)


def _parse_tie_breaker(lines: list[str]) -> tuple[str, str]:
    """Parse the tie breaker section."""
    question = ""
    answer = ""
    for i, line in enumerate(lines):
        if re.match(r"Tie\s*Breaker:", line, re.IGNORECASE):
            question = re.sub(r"Tie\s*Breaker:\s*", "", line).strip()
        elif "price is right" in line.lower():
            continue
        elif question and not answer and line and "make an announcement" not in line.lower():
            answer = line
    return question, answer


def parse_folder(folder: str | Path) -> list[Quiz]:
    """Parse all .docx files in a folder."""
    folder = Path(folder)
    quizzes = []
    for f in sorted(folder.glob("*.docx")):
        if f.name.startswith("~$"):  # skip temp files
            continue
        try:
            quiz = parse_docx(f)
            quizzes.append(quiz)
            print(f"Parsed: {f.name} → Quiz #{quiz.quiz_number} ({quiz.date})")
        except Exception as e:
            print(f"Error parsing {f.name}: {e}")
    return quizzes


if __name__ == "__main__":
    import sys

    path = sys.argv[1] if len(sys.argv) > 1 else "bank"
    p = Path(path)

    if p.is_dir():
        quizzes = parse_folder(p)
        for q in quizzes:
            print(f"\nQuiz #{q.quiz_number} — {q.date}")
            for r in q.rounds:
                if r.round_type == "progressive":
                    print(f"  Round {r.number}: {r.title} ({r.round_type}, {len(r.clues)} clues → {r.progressive_answer})")
                else:
                    print(f"  Round {r.number}: {r.title} ({r.round_type}, {len(r.questions)}Q, {r.points_per_question}pt, joker={'Y' if r.joker_eligible else 'N'})")
            for mg in q.mini_games:
                print(f"  Mini Game {mg.number}: {mg.title}")
            if q.tie_breaker_question:
                print(f"  Tie Breaker: {q.tie_breaker_question[:60]}...")
    else:
        quiz = parse_docx(p)
        print(json.dumps(quiz.to_dict(), indent=2))

File: web/src/lib/parser.ts
import mammoth from "mammoth";
import { extractText } from "unpdf";
import type { Quiz, Round, Question, MiniGame } from "./types";

/**
 * Parse a .docx buffer into a Quiz object.
 */
export async function parseDocx(buffer: Buffer): Promise<Quiz> {
  const result = await mammoth.extractRawText({ buffer });
  return parseQuizText(result.value);
}

/**
 * Parse a .pdf buffer into a Quiz object.
 */
export async function parsePdf(buffer: Buffer): Promise<Quiz> {
  const { text } = await extractText(new Uint8Array(buffer));
  // text is an array of strings (one per page), join them
  const fullText = Array.isArray(text) ? text.join("\n") : String(text);
  return parseQuizText(fullText);
}

/**
 * Parse a buffer of either format based on file extension.
 */
export async function parseFile(
  buffer: Buffer,
  filename: string
): Promise<Quiz> {
  const ext = filename.split(".").pop()?.toLowerCase();
  if (ext === "pdf") return parsePdf(buffer);
  return parseDocx(buffer);
}

/**
 * Core parser: takes raw text content and extracts quiz structure.
 */
export function parseQuizText(rawText: string): Quiz {
  const paragraphs = rawText.split("\n").map((l) => l.trim());

  // Extract quiz number and date from first non-empty line
  let quizNumber = 0;
  let date = "";
  for (const p of paragraphs) {
    // Match "Pub Quiz N – Date" or just "Quiz N – Date"
    const m = p.match(/^(?:Pub\s+)?Quiz\s+(\d+)\s*[–\-]\s*(.+)/i);
    if (m) {
      quizNumber = parseInt(m[1]);
      date = m[2].trim();
      break;
    }
  }

  const quiz: Quiz = {
    quiz_number: quizNumber,
    date,
    rounds: [],
    mini_games: [],
    tie_breaker_question: "",
    tie_breaker_answer: "",
  };

  // Find all section starts
  type Section = [
    "round" | "minigame" | "tiebreaker",
    number,
    number,
    string,
  ];
  const sections: Section[] = [];

  for (let i = 0; i < paragraphs.length; i++) {
    const p = paragraphs[i];
    const rm = p.match(/^Round\s+(\d+)\s*[–\-]\s*(.+)/);
    if (rm) {
      sections.push(["round", i, parseInt(rm[1]), rm[2].trim()]);
    } else if (/^Mini Game \d+/i.test(p)) {
      sections.push(["minigame", i, 0, p]);
    } else if (/^Tie\s*Breaker/i.test(p)) {
      sections.push(["tiebreaker", i, 0, ""]);
    }
  }

  // Deduplicate rounds — keep the occurrence with the most content
  const roundOccurrences: Record<number, number[]> = {};
  for (let i = 0; i < sections.length; i++) {
    if (sections[i][0] === "round") {
      const roundNum = sections[i][2];
      if (!roundOccurrences[roundNum]) roundOccurrences[roundNum] = [];
      roundOccurrences[roundNum].push(i);
    }
  }

  const skipIndices = new Set<number>();
  for (const indices of Object.values(roundOccurrences)) {
    if (indices.length <= 1) continue;
    // For each duplicate round, keep the one with the most lines of content
    let bestIdx = indices[0];
    let bestLines = 0;
    for (const idx of indices) {
      const startLine = sections[idx][1];
      const nextSectionLine =
        idx + 1 < sections.length
          ? sections[idx + 1][1]
          : paragraphs.length;
      const lineCount = nextSectionLine - startLine;
      if (lineCount > bestLines) {
        bestLines = lineCount;
        bestIdx = idx;
      }
    }
    for (const idx of indices) {
      if (idx !== bestIdx) skipIndices.add(idx);
    }
  }

  const deduped = sections.filter((_, i) => !skipIndices.has(i));

  // Process each section
  for (let idx = 0; idx < deduped.length; idx++) {
    const section = deduped[idx];
    const startLine = section[1];
    const endLine =
      idx + 1 < deduped.length ? deduped[idx + 1][1] : paragraphs.length;
    const sectionText = paragraphs.slice(startLine, endLine);

    if (section[0] === "round") {
      quiz.rounds.push(parseRound(section[2], section[3], sectionText));
    } else if (section[0] === "minigame") {
      quiz.mini_games.push(parseMiniGame(sectionText));
    } else if (section[0] === "tiebreaker") {
      const [q, a] = parseTieBreaker(sectionText);
      quiz.tie_breaker_question = q;
      quiz.tie_breaker_answer = a;
    }
  }

  return quiz;
}

function parseRound(
  roundNum: number,
  title: string,
  lines: string[]
): Round {
  const isVideo =
    title.toLowerCase().includes("video") ||
    lines
      .slice(0, 5)
      .some(
        (l) =>
          l.toLowerCase().includes("video") &&
          l.toLowerCase().includes("montage")
      );
  const isProgressive = lines.some((l) => /^\d+\s+POINTS?:/i.test(l));

  let points = 1;
  let joker = true;
  let description = "";

  for (const l of lines.slice(0, 10)) {
    const pm = l.match(/^(\d+)\s+Points?\s+Per\s+Correct\s+Answer/i);
    if (pm) points = parseInt(pm[1]);
    if (
      l.toLowerCase().includes("joker is not in play") ||
      l.toLowerCase().includes("no joker")
    )
      joker = false;
    if (
      l &&
      !/^Round\s+\d+/i.test(l) &&
      !l.toLowerCase().includes("point") &&
      !l.toLowerCase().includes("joker") &&
      !/^\d+\./.test(l) &&
      !l.toLowerCase().includes("video montage") &&
      !l.toLowerCase().includes("youtube")
    ) {
      if (!description && l.length > 20) description = l;
    }
  }

  if (isProgressive) {
    return parseProgressiveRound(
      roundNum,
      title,
      lines,
      points,
      joker,
      description
    );
  }

  const roundType = isVideo ? "video" : "standard";
  const rnd: Round = {
    number: roundNum,
    title,
    round_type: roundType,
    points_per_question: points,
    joker_eligible: joker,
    theme_description: description,
    questions: [],
    clues: [],
    progressive_answer: "",
  };

  // Extract questions and answers
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    const qm = line.match(/^(\d+)\.\s+(.+)/);
    if (qm) {
      const qNum = parseInt(qm[1]);
      const qText = qm[2].trim();
      const isInternet =
        qText.toLowerCase().includes("internet") &&
        qText.toLowerCase().includes("question");

      const choices: string[] = [];
      let answer = "";
      let j = i + 1;

      while (j < lines.length) {
        const nextLine = lines[j];
        if (!nextLine) {
          j++;
          continue;
        }
        if (/^[A-D]\.\s+/.test(nextLine)) {
          choices.push(nextLine);
          if (
            nextLine.toLowerCase().includes("correct") ||
            nextLine.includes("✅")
          ) {
            answer = nextLine.replace(/^[A-D]\.\s+/, "");
            answer = answer
              .replace(/\s*\(?\s*✅?\s*Correct\s*\)?\s*/i, "")
              .trim();
          }
          j++;
        } else if (/^\d+\.\s+/.test(nextLine)) {
          break;
        } else if (
          nextLine.startsWith("Please turn") ||
          nextLine.startsWith("Please ") ||
          nextLine.startsWith("Round ")
        ) {
          break;
        } else {
          if (!answer && !/^[A-D]\.\s+/.test(nextLine)) {
            answer = nextLine;
          }
          j++;
          break;
        }
      }

      // For multiple choice without explicit correct marker
      if (choices.length && !answer) {
        while (j < lines.length && !lines[j]) j++;
        if (j < lines.length && !/^\d+\.\s+/.test(lines[j])) {
          answer = lines[j];
        }
      }

      rnd.questions.push({
        number: qNum,
        text: qText,
        answer,
        choices,
        is_internet_only: isInternet,
      });
      i = j;
    } else {
      i++;
    }
  }

  return rnd;
}

function parseProgressiveRound(
  roundNum: number,
  title: string,
  lines: string[],
  _points: number,
  joker: boolean,
  description: string
): Round {
  const rnd: Round = {
    number: roundNum,
    title,
    round_type: "progressive",
    points_per_question: 0,
    joker_eligible: joker,
    theme_description: description,
    questions: [],
    clues: [],
    progressive_answer: "",
  };

  for (const line of lines) {
    const m = line.match(/^(\d+)\s+POINTS?:\s*(.+)/i);
    if (m) {
      rnd.clues.push({ points: parseInt(m[1]), text: m[2].trim() });
    } else if (/^ANSWER:\s*/i.test(line)) {
      rnd.progressive_answer = line.replace(/^ANSWER:\s*/i, "").trim();
    }
  }

  // Fallback: grab first non-boilerplate line after last clue
  if (!rnd.progressive_answer && rnd.clues.length) {
    let lastClueIdx = 0;
    for (let i = 0; i < lines.length; i++) {
      if (/^\d+\s+POINTS?:/i.test(lines[i])) lastClueIdx = i;
    }
    for (let j = lastClueIdx + 1; j < lines.length; j++) {
      const line = lines[j];
      if (
        line &&
        !line.startsWith("Please") &&
        !/^Round\s+\d+/i.test(line) &&
        !/^\d+\s+Points?\s+Per/i.test(line) &&
        !/^Mini Game/i.test(line) &&
        !/^Tie\s*Breaker/i.test(line)
      ) {
        rnd.progressive_answer = line;
        break;
      }
    }
  }

  return rnd;
}

function parseMiniGame(lines: string[]): MiniGame {
  const header = lines[0] || "";
  const numMatch = header.match(/^Mini Game (\d+)/i);
  const num = numMatch ? parseInt(numMatch[1]) : 0;
  const title = header.replace(/^Mini Game \d+\s*[–\-]\s*/, "").trim();
  const desc = lines
    .slice(1)
    .filter((l) => l)
    .join("\n");
  return { number: num, title, description: desc };
}

function parseTieBreaker(lines: string[]): [string, string] {
  let question = "";
  let answer = "";
  for (const line of lines) {
    if (/^Tie\s*Breaker:\s*/i.test(line)) {
      question = line.replace(/^Tie\s*Breaker:\s*/i, "").trim();
    } else if (line.toLowerCase().includes("price is right")) {
      continue;
    } else if (
      question &&
      !answer &&
      line &&
      !line.toLowerCase().includes("make an announcement")
    ) {
      answer = line;
    }
  }
  return [question, answer];
}
